Zend/zend_string.c

```diff
@@ -372,11 +372,57 @@ ZEND_API void zend_interned_strings_switch_storage(bool request)
 # define I_REPLACE_SONAME_FNNAME_ZU(soname, fnname) _vgr00000ZU_ ## soname ## _ ## fnname
 #endif
 
-ZEND_API bool ZEND_FASTCALL I_REPLACE_SONAME_FNNAME_ZU(NONE,zend_string_equal_val)(zend_string *s1, zend_string *s2)
+#define VALGRIND_REPLACEMENT_STRING_EQUAL_VAL_MEMCMP_WRAPPER I_REPLACE_SONAME_FNNAME_ZU(NONE,zend_string_equal_val_memcmp_wrapper)
+#define VALGRIND_REPLACEMENT_STRING_EQUAL_VAL I_REPLACE_SONAME_FNNAME_ZU(NONE,zend_string_equal_val)
+#define STR1(X) #X
+#define STR(X) STR1(X)
+
+bool VALGRIND_REPLACEMENT_STRING_EQUAL_VAL_MEMCMP_WRAPPER(zend_string *s1, zend_string *s2)
 {
 	return !memcmp(ZSTR_VAL(s1), ZSTR_VAL(s2), ZSTR_LEN(s1));
 }
 
+#if defined(__GNUC__) && defined(__x86_64__) && !defined(__ILP32__) && !defined(__MACH__)
+/*
+ * See https://github.com/php/php-src/issues/9068
+ * We need to preserve the scratch registers because we cannot know what
+ * the compiler assumes to be unchanged by the original
+ * zend_string_equal_val implementation.
+ * We can safely use push and pop because we're not in leaf function.
+ * We're using a function defined using inline assembly so that the compiler does not
+ * mess with our store and restore procedure.
+ */
+__asm__ (
+	".globl " STR(VALGRIND_REPLACEMENT_STRING_EQUAL_VAL) "\n\t"
+	".type " STR(VALGRIND_REPLACEMENT_STRING_EQUAL_VAL) ", @function\n\t"
```

Comment:
@nielsdos you might not need much adaptation for macOs, indeed as you said the `.type` and `.size` directives are unsupported.

Reply:
So you think it suffices to guard these with an ifdef? I'm asking because I can't test on macOS myself :sweat_smile:

Reply:
I did rapid local changes on my mac in x86 64 emulation mode, seems to work fine.